src/newgrf.cpp

```diff
@@ -9161,6 +9161,13 @@ static void CalculateRefitMasks()
 				ei->cargo_type = (CargoID)FindFirstBit(ei->refit_mask);
 			}
 		}
+		if (!IsValidCargoID(ei->cargo_type) && e->type == VEH_TRAIN && e->u.rail.railveh_type != RAILVEH_WAGON && e->u.rail.capacity == 0) {
+			/* For train engines which do not carry cargo it does not matter if their cargo type is invalid, fallback to
+			 * the first available instead, if the cargo type has not been changed. */
+			if (GetActiveCargoLabel(ei->cargo_label) != CT_INVALID) {
```

Comment:
Is this check still needed? You already concluded the cargoID is not valid. Would this ever return anything other than `CT_INVALID`?

Reply:
This check is to ensure this change only applies if the default cargo type (stored as a label) has not been changed. 

This value is set to CT_INVALID only when cargo_type has been changed by NewGRF.

Reply:
I've updated the comment to document this slightly better.